collection-scripts/gather-sosreports

```diff
@@ -0,0 +1,184 @@
+#!/bin/bash
+
+BASE_COLLECTION_PATH="/must-gather"
+SOSREPORT_DUMP_PATH=${OUT:-"${BASE_COLLECTION_PATH}/node_sosreports"}
+# 20 minutes timeout for a sosreport pod should be reasonablef
+POD_TIMEOUT=1200
+# Do not run a sosreport if the minimum availble GB is not met on /var/tmp
+MIN_AVAILBLE_GB=20
+
+# Gather a sosreport on a specific node
+# Parameters:
+# $1 - node name: argument to oc debug node in format node/<name>
+# $2 - registry:  URL to custom registry for image overwrite of toolbox image
+# $3 - image:     image name for image overwrite of toolbox image
+function gather_sosreport() {
+	local node=$1
+	local registry=$2
+	local image=$3
+
+	# Because we have to chroot, sudo and start a toolbox container
+	# it is easier to run this in a script which is then copied to
+	# the debug pod
+	local tmp_file=$(mktemp)
+	cat <<'EOF' > $tmp_file
+#!/bin/bash
+
+chroot /host /usr/bin/sudo --preserve-env /bin/bash <<'EOT'
+# Defaults first
+REGISTRY=registry.redhat.io
+IMAGE=rhel8/support-tools
+TOOLBOX_NAME=toolbox-"$(whoami)-sosreport-$(uuidgen | cut -b-8)"
+TOOLBOXRC="/root/.toolboxrc"
+
+# Second if .toolboxrc file exists, take its values
+if [ -f "${TOOLBOXRC}" ]; then
+    echo ".toolboxrc file detected, overriding defaults..."
+    source "${TOOLBOXRC}"
+fi
+
+# Third if user overrides are set, use those
+USER_REGISTRY=""
+USER_IMAGE=""
+if [ "${USER_REGISTRY}" != "" ]; then
+    REGISTRY="${USER_REGISTRY}"
+fi
+if [ "${USER_IMAGE}" != "" ]; then
+    IMAGE="${USER_IMAGE}"
+fi
+
+# Now, concat registry and image
+TOOLBOX_IMAGE="${REGISTRY}"/"${IMAGE}"
+
+podman pull --authfile /var/lib/kubelet/config.json "$TOOLBOX_IMAGE"
+
+podman run \
+  --rm \
+  --hostname $(hostname) \
+  --name "$TOOLBOX_NAME" \
+  --privileged \
+  --net=host \
+  --pid=host \
+  --ipc=host \
+  -e HOST=/host \
+  -e NAME="$TOOLBOX_NAME" \
+  -e IMAGE="$IMAGE" \
+  --security-opt label=disable \
+  --volume /run:/run \
+  --volume /var/log:/var/log \
+  --volume /etc/machine-id:/etc/machine-id \
+  --volume /etc/localtime:/etc/localtime \
+  --volume /:/host \
+   "$TOOLBOX_IMAGE" \
+  sosreport -k crio.all=on -k crio.logs=on --batch --all-logs 2>/dev/null | tee /tmp/must_gather_sosreport.log
```

Comment:
One thing to consider here is varying versions of `sos` available on each node/image, and the ability to have users pass/change options as needed.

For `sos collect` this is handled on a per-node basis, meaning `collect` is aware of the capabilities of sos on each individual node. If plugins, options, etc... are passed by the user but aren't supported by particular nodes, the options are trimmed for those nodes alone to still allow a successful command. Invalid options or more specifically invalid plugin names will otherwise result in a failed collection.

Then you also need to consider how do you allow users to disable arbitrary collections (e.g. `-n $plugin_name` or `--skip-commands=$some_cmd`) that may be causing issues for otherwise "standard" sos runs?

Tangentially, I would very strongly recommend dropping `--all-logs` as a default as that removes all size limiting which will not only balloon the file archive size on disk, but will also significantly increase the memory usage of sos. Further, the combination of `-k {docker|podman|crio}.all=on -k {docker|podman|crio}.logs=on` is something I have been advocating against for some time as a default configuration as that means "give me all logs (everything the container sent to stdout) for every container that exists on this host even if they aren't currently running". Combined with `--all-logs` this can quickly become extremely expensive in terms of disk and memory usage.

Reply:
From a "field" perspective of almost 3 years in Support and 2 more as an OCP customer, OCP troubleshooting (specially OCP4) implies looking what happens in a number of containers again and again. 

I end up **always** requesting all the logs from all the containers and I do need them 95% of the times (not an exaggeration). Maybe I don't need every single container logs, but asking sosreport and then logs of the relevant containers is usually not an acceptable approach, because the container may have vanished or its logs may have been rotated by the time you ask them separately.

On the other hand, bigger memory consumption and disk stress have been a problem only in very very few scenarios, where the node was so broken that I also needed to request other usually working sosreport modules to be disable with `-n`, i.e. these were very exceptional situations. 

And in what regards sosreport size... must-gathers are usually much bigger than a single sosreport (even after audit logs are not collected by default), so our customers are kind of used to share large data sets already.

So if these options are not made default, I foresee myself and other colleagues just always asking for them, because we are going to need them. If they are not provided at all, maybe we would need to build an alternative and not use this, because in short: We do need them most of the times.